### ISSUE TITLE
[FSDP] why I cannot align the loss exactly in equivalent situation?

### ISSUE THREAD

**Issue by vincentccc** [pytorch/pytorch#171874]
I use mix precision in my training script, such like
```
fsdp_module 1 (bf16 noshard or shard grad)
     module 1.1
     fsdp_module 1.2 (which is FP32 noshard)
```

I compare the grad of `fsdp_module 1.2 `after backward in 2 rank respectively, they are bit-wise same, but after allreduce(save the grad in optimizer step) it shows different. I use torch.profile to ensure they both use FP32 nccl allreduce, I‘m confused about why ? And how can I further debug it?

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci
